File: src/main/java/core/procedures/system/ToString.java
package core.procedures.system;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.writer.Writer;

public class ToString extends AFn {

  public ToString() {
    super(new FnArgsBuilder().min(0).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "->string";
  }

  @Override
  public CharSequence apply(Object... args) {
    if (args.length == 0) {
      return "";
    }
    if (args.length == 1) {
      return str(args[0]);
    }
    StringBuilder sb = new StringBuilder();
    for (Object arg : args) {
      sb.append(str(arg));
    }
    return sb.toString();
  }

  private CharSequence str(Object obj) {
    if (obj == null) {
      return "";
    }
    if (obj instanceof Character) {
      return obj.toString();
    }
    if (obj instanceof CharSequence) {
      return (CharSequence) obj;
    }
    return Writer.write(obj);
  }
}

File: src/main/java/core/procedures/system/ToString.kt
package core.procedures.system

import core.procedures.AFn
import core.Writer

open class ToString : AFn<Any?, CharSequence>(name = "->string", isPure = true) {

    override operator fun invoke(args: Array<out Any?>) = args.joinToString(separator = "") { str(it) }

    private fun str(obj: Any?) = when (obj) {
        null            -> ""
        is Char         -> obj.toString()
        is CharSequence -> obj
        else            -> Writer.write(obj)
    }
}
